Include/pythread.h

```diff
@@ -17,7 +17,37 @@ typedef enum PyLockStatus {
 
 PyAPI_FUNC(void) PyThread_init_thread(void);
 PyAPI_FUNC(unsigned long) PyThread_start_new_thread(void (*)(void *), void *);
-PyAPI_FUNC(void) _Py_NO_RETURN PyThread_exit_thread(void);
+/* Terminates the current thread.
+ *
+ * WARNING: This function is only safe to call if all functions in the full call
+ * stack are written to safely allow it.  Additionally, the behavior is
+ * platform-dependent.  This function should be avoided, and is no longer called
+ * by Python itself.  It is retained only for compatibility with existing C
+ * extension code.
+ *
+ * With pthreads, calls `pthread_exit` which attempts to unwind the stack and
+ * call C++ destructors.  If a `noexcept` function is reached, the program is
+ * terminated.
+ *
+ * On Windows, calls `_endthreadex` which kills the thread without calling C++
+ * destructors.
+ *
+ * In either case there is a risk of invalid references remaining to data on the
+ * thread stack.
+ */
+Py_DEPRECATED(3.12) PyAPI_FUNC(void) _Py_NO_RETURN PyThread_exit_thread(void);
```

Comment:
Since `PyThread_exit_thread` is part of the public API of CPython and has been for some time, I wanted to preserve the existing behavior.  To know what behavior is appropriate for each use of it would require analyzing each call site.  By marking it deprecated, users will receive a warning and can do that work themselves.

For the existing uses in CPython itself when failing to acquire the GIL, hanging the thread seems to be the best available behavior, but for unknown external call sites it is impossible to know what the best behavior is, but I suspect that hanging the thread would not be the desired behavior in most external uses of this function.

Reply:
the signal stuff is in another comment thread: don't block signals - there is no reason to bother doing that.  i left a comment on that part of the code separately.

otherwise I agree with this dangerous API having a deprecation marker and the new hang API being created rather than transforming the dangerous one into a silent hang.